File: supabase/functions/_shared/brokeMode.ts
// وضع الطوارئ — منطق مشترك بين zad-brain وzad-core-intelligence.

/**
 * وضع الطوارئ «مفلس باقي الشهر» (20260914009000). صافية: خطة الأيام الباقية.
 * `cashLeft` = اللي العميل قاله بنفسه، وإلا المتاح من zad_budget_state لو السقف متأكد.
 * الأيام من days_left (أقل حاجة يوم)، والنهاية من cycle_end (حصري) أو الأيام الباقية.
 */
export function brokeModePlan(input: {
  cashLeft: number | null;
  available: number | null;
  limitConfirmed: boolean;
  daysLeft: number | null;
  cycleEnd: string | null;
  nowMs: number;
}): { cash_left: number | null; daily_cap: number | null; days_left: number; ends_at: string } {
  const days = Math.max(1, Math.min(45, Math.round(Number(input.daysLeft) || 0) || 1));
  const cash = input.cashLeft !== null && Number.isFinite(input.cashLeft)
    ? Math.max(0, input.cashLeft)
    : input.limitConfirmed && typeof input.available === "number" && Number.isFinite(input.available)
      ? Math.max(0, input.available)
      : null;
  const byCycle = input.cycleEnd && /^\d{4}-\d{2}-\d{2}$/.test(input.cycleEnd)
    ? Date.parse(`${input.cycleEnd}T00:00:00Z`)
    : NaN;
  const endMs = Number.isFinite(byCycle) && byCycle > input.nowMs + 3_600_000
    ? byCycle
    : input.nowMs + days * 86_400_000;
  return {
    cash_left: cash === null ? null : Math.round(cash * 100) / 100,
    daily_cap: cash === null ? null : Math.floor(cash / days),
    days_left: days,
    ends_at: new Date(endMs).toISOString(),
  };
}

/** الوضع شغال دلوقتي؟ صف قديم خلص وقته = مش شغال حتى لو ended_at فاضي. */
export function isBrokeModeActive(row: { ends_at?: string | null; ended_at?: string | null } | null, nowMs: number): boolean {
  if (!row || row.ended_at) return false;
  const end = Date.parse(String(row.ends_at ?? ""));
  return Number.isFinite(end) && end > nowMs;
}

/** وصفات الطوارئ: مفيش ولا صنف يتشرى. البهارات والملح والزيت والمية مش «شراء». */
const PANTRY_STAPLES = ["ملح", "فلفل", "زيت", "مية", "ماء", "بهارات", "سكر", "كمون"];
export function recipeNeedsNoShopping(missing: unknown): boolean {
  if (!Array.isArray(missing)) return true;
  return missing.every((m) => {
    const name = String(m ?? "").trim();
    return name === "" || PANTRY_STAPLES.some((s) => name.includes(s));
  });
}

File: app/src/main/java/com/example/data/BrokeMode.kt
package com.example.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import kotlin.math.floor

/**
 * وضع الطوارئ «مفلس باقي الشهر» (`zad_broke_mode`، ميجريشن 20260914009000).
 *
 * العميل بيقول «أنا مفلس» للعقل (شات/مكالمة/تليجرام → set_broke_mode) أو بيفعّله من شاشة
 * الميزانية. طول ما هو شغال: مصروف اليوم = اللي معاه ÷ الأيام الباقية، اقتراحات الشراء
 * بتختفي، والوصفات من اللي في البيت بس.
 */
@Serializable
data class ZadBrokeMode(
    @SerialName("user_id") val userId: String? = null,
    @SerialName("started_at") val startedAt: String? = null,
    @SerialName("ends_at") val endsAt: String,
    @SerialName("ended_at") val endedAt: String? = null,
    @SerialName("cash_left") val cashLeft: Double? = null,
    @SerialName("daily_cap") val dailyCap: Double? = null,
    val currency: String? = null,
    val source: String = "app",
)

private fun parseInstant(raw: String?): Instant? = raw?.let {
    runCatching { Instant.parse(it) }.getOrNull()
        ?: runCatching { java.time.OffsetDateTime.parse(it).toInstant() }.getOrNull()
}

/** شغال = مااتقفلش ولسه وقته ماخلصش. صف قديم خلص وقته = مش شغال. */
fun ZadBrokeMode?.isActive(now: Instant = Instant.now()): Boolean {
    if (this == null || endedAt != null) return false
    val end = parseInstant(endsAt) ?: return false
    return end.isAfter(now)
}

object BrokeModeMath {
    data class Plan(val cashLeft: Double?, val dailyCap: Double?, val daysLeft: Int, val endsAt: Instant)

    /** نفس `brokeModePlan` في `_shared/brokeMode.ts` — التطبيق والعقل لازم يطلعوا نفس الرقم. */
    fun plan(
        cashLeft: Double?,
        available: Double?,
        limitConfirmed: Boolean,
        daysLeft: Int,
        cycleEnd: LocalDate?,
        now: Instant,
    ): Plan {
        val days = daysLeft.coerceIn(1, 45)
        val cash = when {
            cashLeft != null && cashLeft.isFinite() -> cashLeft.coerceAtLeast(0.0)
            limitConfirmed && available != null && available.isFinite() -> available.coerceAtLeast(0.0)
            else -> null
        }
        val byCycle = cycleEnd?.atStartOfDay()?.toInstant(ZoneOffset.UTC)
        val end = if (byCycle != null && byCycle.isAfter(now.plusSeconds(3600))) byCycle
        else now.plusSeconds(days * 86_400L)
        return Plan(
            cashLeft = cash?.let { kotlin.math.round(it * 100) / 100 },
            dailyCap = cash?.let { floor(it / days) },
            daysLeft = days,
            endsAt = end,
        )
    }

    /** بيانات مطابقة (مش نصوص عرض): حاجات في كل مطبخ مش بتتحسب «شراء». نفس القايمة في السيرفر. */
    private val PANTRY_STAPLES = listOf("ملح", "فلفل", "زيت", "مية", "ماء", "بهارات", "سكر", "كمون")

    fun recipeNeedsNoShopping(missing: List<String>): Boolean =
        missing.all { m -> m.isBlank() || PANTRY_STAPLES.any { m.contains(it) } }
}
